lib/sigstore/signer.rb

```diff
@@ -0,0 +1,246 @@
+# frozen_string_literal: true
+
+# Copyright 2024 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require_relative "internal/util"
+require_relative "internal/x509"
+require_relative "models"
+require_relative "policy"
+require_relative "verifier"
+
+module Sigstore
+  class Signer
+    include Loggable
+
+    def initialize(jwt:, trusted_root:)
+      @jwt = jwt
+      @trusted_root = trusted_root
+    end
+
+    def sign(payload)
+      # 2) generate a keypair
+      keypair = generate_keypair
+      # 3) generate a CreateSigningCertificateRequest
+      csr = generate_csr(keypair)
+      # 4) get a cert chain from fulcio
+      chain = fetch_cert(csr)
+      # 5) verify returned cert chain
+      leaf, _chain = verify_chain(chain)
+      # 6) sign the payload
+      signature = sign_payload(payload, keypair)
+      # 7) send hash of signature to timestamping service
+      timestamp_verification_data = submit_signature_hash_to_timstamping_service(signature)
+      # 8) submit signed metadata to transparency service
+      tlog_entries = submit_signed_metadata_to_transparency_service(signature, leaf,
+                                                                    OpenSSL::Digest("SHA256").digest(payload))
+      # 9) perform verification
+
+      verification_material = Bundle::V1::VerificationMaterial.decode_json_hash(
+        {
+          "x509CertificateChain" => {
+            "certificates" => chain[0..-2].map { |c| { "rawBytes" => Internal::Util.base64_encode(c.to_der) } }
+          },
+          "tlogEntries" => tlog_entries,
+          "timestampVerificationData" => timestamp_verification_data
+        },
+        registry: REGISTRY
+      )
+      verify(payload, verification_material, signature)
+      # TODO: return a bundle
+      [signature, leaf, verification_material]
+    end
+
+    private
+
+    def generate_keypair
+      # TODO: check if the type of key matters?
```

Comment:
- [ ] todo: key type

Reply:
The type of key *should* ultimately be configurable, but most other clients default to P-256 (i.e. prime256v1) and not all are configurable yet. So the current default is reasonable here and IMO configurability should probably be a follow-up.